[MRG+2] GSoC Final : Dirichlet Gaussian Mixture

tguillemot opened the pull request:
This closes #7377, closes #7115, closes #2473, closes #2454, closes #1764 and closes #1637.

This is the last PR to remove completely the old GMM classes.

Here, you'll find the `DirichletGaussianMixture` class with the doc, examples and tests.

~~It will be easier to review when #6651 will be merged (and also there will be no conflits).~~

<del>I have removed the example `plot_gmm_sin.py` because it's wasn't showing the properties of DPGMM correctly for me (it modifies the `covariance_type` through the experiment to obtain better results).</del>


Instead of that I prefer introduce a similar exemple as the one I have introduce for the `BayesianGaussianMixture`.
![dpgmm](https://cloud.githubusercontent.com/assets/17002898/18122406/a0f33a34-6f69-11e6-8a63-512f315c1207.png)

tguillemot committed:
Add DirichletGaussianMixture + doc + example + tests.

tguillemot committed:
Update Deprecation Warning and correct docstring.

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -226,34 +227,41 @@ concentration parameter.
+
 
+The examples above compare Gaussian mixtures models with fixed number of
+components, to the Dirichlet Gaussian Mixtures models. **On the left** the  GMM
```
GMM -> Gaussian mixtures

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -226,34 +227,41 @@ concentration parameter.
 
-.. |plot_gmm_sin| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_sin_001.png
-   :target: ../auto_examples/mixture/plot_gmm_sin.html
+.. |plot_gmm_sin| image:: ../auto_examples/mixture/images/sphx_glr_plot_dirichlet_process_mixture_001.png
```
`plot_gmm_sin` does not exist anymore

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -226,34 +227,41 @@ concentration parameter.
+components, to the Dirichlet Gaussian Mixtures models. **On the left** the  GMM
+is fitted with 5 components on a dataset composed of 2 clusters. We can see that
+the Dirichlet Gaussian Mixtures is able to limit itself to only 2 components
+whereas the GMM fits the data fit too many components. Note that with very
```
GMM/DPGMM here and also above in the rest of the file

TomDLT commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+
+Plot the resulting ellipsoids of a mixture of three Gaussians with
+a Dirichlet Process Gaussian Mixture for three different values of the prior
+the beta concentration.
```
the beta concentration prior?

TomDLT commented on sklearn/mixture/dirichlet_gaussian_mixture.py:
```diff
@@ -0,0 +1,333 @@
+    beta_concentration_beta_ : array-like, shape (`n_components`, )
+        The beta concentration parameters of the Beta distribution of each
+        component.
+
```
two lines

TomDLT commented on sklearn/mixture/dirichlet_gaussian_mixture.py:
```diff
@@ -0,0 +1,333 @@
+                 mean_precision_prior=None, mean_prior=None,
+                 degrees_of_freedom_prior=None, covariance_prior=None,
+                 random_state=None, warm_start=False, verbose=0,
+                 verbose_interval=20):
```
verbose_interval no in the docstring

TomDLT commented on sklearn/mixture/dirichlet_gaussian_mixture.py:
```diff
@@ -0,0 +1,333 @@
+        X : array-like, shape (n_samples, n_features)
+
+        log-resp : array, shape (n_samples, n_components)
+
```
as in `BayesianGaussianMixture`:
```
        log_resp : array, shape (n_samples, n_components)
            Logarithm of the posterior probabilities (or responsibilities) of
            the point of each sample in X.

        log_prob_norm : float
            Logarithm of the probability of each sample in X.
```

TomDLT commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+# Parameters
+random_state = 2
+n_components, n_features = 3, 2
+colors = np.array(['mediumseagreen', 'royalblue', 'r', 'gold',
```
You modified the colors. Are you sure its colorblind compatible ?
cf #5576

TomDLT replied:
also in BayesianGaussianMixture example

tguillemot commented:
@TomDLT Thanks for this first round of review.

ogrisel commented:
I am in favor of keeping the "sin" example (adapted to use the new class). While I agree that it is a weird and artificial test dataset, I also appreciate the fact that:

- it's bad to break sacred links on the Holy Web,
- it's good to show that models can in practice still be used (and be somewhat useful) even if the data violate the assumption of the underlying generative model.

tguillemot committed:
Add the `gmm_sin.py`.

tguillemot commented:
Ok fair enough. So, I've modified the example to plot something that make sense (change only the beta concentration prior and not the `covariance_type` ).
![gmm_sin](https://cloud.githubusercontent.com/assets/17002898/18168493/4ab3da14-7056-11e6-936a-f2e7e9ce6614.png)

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,68 @@ Cons
 .. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
    :target: ../auto_examples/mixture/plot_gmm.html
-   :scale: 48%
+   :scale: 31%
```
I guess the [rendered page](https://27-50829930-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/mixture.html#dirichlet-process-infinite-gaussian-mixtures) is not as intended.

tguillemot replied:
I was wondering what will be the result and it's ugly :).
I'll change that.

TomDLT commented:
This seems pretty clean to me

tguillemot commented:
[This is the new doc](https://32-50829930-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/mixture.html).

@agramfort @ogrisel Do you think we can merge that for 0.18 ???

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -93,14 +93,14 @@ Cons
+The BIC criterion can be used to select the number of components in a Gaussian
+Mixture in an efficient way. In theory, it recovers the true number of
+components only in the asymptotic regime (i.e. if much data is available). Note
+that using a :ref:`DirichletGaussianMixture <dpgmm>` avoids the specification of
```
Since this is not a class ref, could we have this as `"Dirichlet Gaussian Mixture"`?

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
+Dirichlet Process: Infinite Gaussian mixtures
+==============================================================
+
+The :class:`DirichletGaussianMixture` object implements a variant of the
```
~~object~~

ogrisel replied:
It should be:

 ```
The :class:`DirichletGaussianMixture` class implements...
```

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
+   :align: center
+   :scale: 70%
+
+Here, a classical Gaussian mixtures is fitted with 5 components on a dataset
```
mixture?

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
 
-.. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
+
+The examples bellow compare Gaussian mixtures models with fixed number of
```
mixtures --> mixture

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
-.. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
+
+The examples bellow compare Gaussian mixtures models with fixed number of
+components, to the Dirichlet Gaussian mixtures models.
```
mixtures --> mixture

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
+Here, a classical Gaussian mixtures is fitted with 5 components on a dataset
+composed of 2 clusters. We can see that the Dirichlet Gaussian mixtures is able
+to limit itself to only 2 components whereas the Gaussian mixtures fits the data
+fit too many components. Note that with very little observations, the Dirichlet
```
fit too --> with too

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
-data.
+The previous figure presents the resulting clusters computed by the Dirichlet
+Gaussian mixtures for different values of `beta_concentration_prior`. The
+`beta_concentration_prior` is directly linked to the number of clusters
```
double backticks here too...

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
+Gaussian mixtures for different values of `beta_concentration_prior`. The
+`beta_concentration_prior` is directly linked to the number of clusters
+obtained. As for the Variational Bayesian Gaussian Mixtures, small value of this
+parameters will lead to some mixture components while high value leads more
```
leads to

raghavrv replied:
Is this better maybe?

```
The ``beta_concentration_prior`` is directly proportional to the number of clusters obtained. In Variational Bayesian Gaussian Mixture, small values of ``beta_concentration_prior`` leads to fewer components and higher values lead to more components but is more stable as each component is activated only if necessary.
```

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,73 @@ Cons
 
-Pros and cons of class :class:`DPGMM`: Dirichlet process mixture model
-----------------------------------------------------------------------
+Pros and cons of class :class:`DirichletGaussianMixture`: Dirichlet process mixture model
```
Capital P, M, M?

tguillemot replied:
It's more `Pros and cons of Dirichlet process with class :class:DirichletGaussianMixture`

raghavrv commented on doc/modules/mixture.rst:
```diff
@@ -266,9 +290,9 @@ Pros
-   this number needs to be provided. Note however that the DPMM is not
-   a formal model selection procedure, and thus provides no guarantee
-   on the result.
+   this number needs to be provided. Note however that the Dirichlet Gaussian
```
Dirichlet Process Gaussian Mixture

raghavrv commented on examples/mixture/plot_bayesian_gaussian_mixture.py:
```diff
@@ -5,7 +5,7 @@
+dirichlet concentration prior.
 
 For all models, the Variationnal Bayesian Gaussian Mixture adapts its number of
 mixture automatically. The parameter `dirichlet_concentration_prior` has a
```
Double ticks here too..

tguillemot replied:
I will correct all the double ticks. :)

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+"""
+=============================================
+Dirichlet Process Mixture Beta Prior Analysis
```
Gaussian Mixture

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+
+Plot the resulting ellipsoids of a mixture of three Gaussians with
+a Dirichlet Process Gaussian Mixture for three different values of the
+beta concentration prior.
```
```
``beta_concentration_prior``
```

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+a Dirichlet Process Gaussian Mixture for three different values of the
+beta concentration prior.
+
+For all models, the Dirichlet Process Gaussian Mixture adapts its number of
```
all the

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+beta concentration prior.
+
+For all models, the Dirichlet Process Gaussian Mixture adapts its number of
+mixture automatically. The parameter `beta_concentration_prior` has a
```
components rather than mixtures for consistency maybe? (ambivalent)

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+For all models, the Dirichlet Process Gaussian Mixture adapts its number of
+mixture automatically. The parameter `beta_concentration_prior` has a
+direct link with the resulting number of components. Specifying a high value of
+`beta_concentration_prior` leads more often to uniformly-sized mixture
```
more often leads

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+
+For all models, the Dirichlet Process Gaussian Mixture adapts its number of
+mixture automatically. The parameter `beta_concentration_prior` has a
+direct link with the resulting number of components. Specifying a high value of
```
Specifying higher values

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+mixture automatically. The parameter `beta_concentration_prior` has a
+direct link with the resulting number of components. Specifying a high value of
+`beta_concentration_prior` leads more often to uniformly-sized mixture
+components, while specifying small values will lead to some mixture
```
specifying smaller values

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+direct link with the resulting number of components. Specifying a high value of
+`beta_concentration_prior` leads more often to uniformly-sized mixture
+components, while specifying small values will lead to some mixture
+components getting almost all the points while most mixture components will be
```
will lead to some mixture components getting almost all the points while most of the other mixture components centering on just a few of the remaining points.

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+        v, w = np.linalg.eigh(covars[n][:2, :2])
+        u = w[0] / np.linalg.norm(w[0])
+        angle = np.arctan2(u[1], u[0])
+        angle = 180 * angle / np.pi  # convert to degrees
```
Could you move the comment to above this line?

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+
+def plot_ellipses(ax, weights, means, covars):
+    for n in range(means.shape[0]):
+        v, w = np.linalg.eigh(covars[n][:2, :2])
```
Isn't the notation `w` used to denote the eigen values and `v` for vectors and hence should it be `w, v`?

raghavrv replied:
Or could you use `eig_vals`, `eig_vecs` instead?

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+def plot_ellipses(ax, weights, means, covars):
+    for n in range(means.shape[0]):
+        v, w = np.linalg.eigh(covars[n][:2, :2])
+        u = w[0] / np.linalg.norm(w[0])
```
`u --> unit_eig_vecs`?

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+        angle = np.arctan2(u[1], u[0])
+        angle = 180 * angle / np.pi  # convert to degrees
+        v = 2 * np.sqrt(2) * np.sqrt(v)
+        ell = mpl.patches.Ellipse(means[n, :2], v[0], v[1], 180 + angle)
```
I think `:2` is not needed?

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+        u = w[0] / np.linalg.norm(w[0])
+        angle = np.arctan2(u[1], u[0])
+        angle = 180 * angle / np.pi  # convert to degrees
+        v = 2 * np.sqrt(2) * np.sqrt(v)
```
I'm not sure if this is needed, but do you think it would be useful to add a comment explaining why we do this?

tguillemot replied:
fair enough

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+    estimator.fit(X)
+    ax1.set_title("Dirichlet Process Mixture for "
+                  r"$\beta_0=%.1e$" % beta_concentration_prior)
+    # ax1.axis('equal')
```
scaffold to be removed

raghavrv commented on examples/mixture/plot_dirichlet_process_mixture.py:
```diff
@@ -0,0 +1,115 @@
+
+# Parameters
+random_state = 2
+n_components, n_features = 3, 2
```
In 2 lines?

ogrisel commented:
> @agramfort @ogrisel Do you think we can merge that for 0.18 ???

Yes. If it's ready by Tuesday next week :) I will try to give it another quick review.

tguillemot commented:
@raghavrv is reviewing the PR, once I've corrected that I send you a message.

ogrisel commented:
Could you please combine the following 2 examples into one:

- plot_bayesian_gaussian_mixture.py
- plot_dirichlet_process_mixture.py

You can still create 2 figures by doing calls to `plt.figure()` to avoid having a single very large figure with many subplots.

Those two examples are very similar and that should make it possible to remove redundant code while at the same time make it easier to intuitively contrast the way the 2 concentration priors work.

tguillemot committed:
Correct reviews and examples.

tguillemot commented:
> Could you please combine the following 2 examples into one

Done ! [Here is the new doc](https://34-50829930-gh.circle-artifacts.com/0//home/ubuntu/scikit-learn/doc/_build/html/stable/modules/mixture.html)

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -224,41 +224,43 @@ only needs to specify a loose upper bound on this number and a
 
-.. figure:: ../auto_examples/mixture/images/sphx_glr_plot_dirichlet_process_mixture_001.png
-   :target: ../auto_examples/mixture/plot_dirichlet_process_mixture.html
+.. figure:: ../auto_examples/mixture/images/sphx_glr_plot_concentration_prior_001.png
```
_002

tguillemot commented:
@raghavrv @ogrisel @TomDLT Thanks for your reviews. Do you have something else to correct ?

TomDLT commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
+      :class:`GaussianMixture` and :class:`DirichletGaussianMixture` to fit a
+      sine wave.
+
+    * :ref:`sphx_glr_auto_example_mixture_plot_concentration_prior_mixture.py`
```
`* See :ref: `

TomDLT replied:
And the link is wrong

TomDLT commented:
otherwise LGTM

tguillemot committed:
Correct reviews.

tguillemot commented:
Thanks @TomDLT

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
-.. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
+
+The examples bellow compare Gaussian mixture models with fixed number of
+components, to the Dirichlet Gaussian mixture models.
```
Gaussian mixture models with a  Dirichlet **process** prior

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
+to limit itself to only 2 components whereas the Gaussian mixture fits the data
+fit with too many components. Note that with very little observations, the
+Dirichlet Gaussian Mixture can take a conservative stand, and fit only one
+component.
```
I would rather put the explaining paragraphs before the matching figures.

tguillemot committed:
Modify the doc.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
-parameter of the DPGMM controls the number of components used to fit this
-data.
+The previous figure presents the resulting clusters computed by the Dirichlet
+Gaussian mixture for different values of ``weight_concentration_prior``. The
```
Gaussian mixture with a Dirichlet process prior.

The "process" is important. 

The weights concentration prior of the BayesianGaussianMixture class is already based on Dirichlet distribution while the weights concentration prior of the DirichletGaussianMixture is based on a Dirichlet process.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -161,23 +161,23 @@ much so as to render usage unpractical.
-      of the parameter ``dirichlet_concentration_prior``.
+      of the parameter ``weight_concentration_prior``.
 
 Pros and cons of variational inference with :class:BayesianGaussianMixture
```
The `class` markup is broken because of missing backticks.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -266,9 +292,9 @@ Pros
-   on the result.
+   this number needs to be provided. Note however that the Dirichlet process
+   Gaussian mixture is not a formal model selection procedure, and thus
+   provides no guarantee on the result.
```
One could argue that this is actually a formal model selection procedure but that does not mean that it is that useful in practice ;) I would rephrase this last sentence to the following:

Note however that the "ideal" number of active components is very application specific and is typically ill-defined in a data exploration setting. The effective number of active components often depends a lot on the value for the weights concentration prior.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -189,8 +189,8 @@ Pros
+``n_components`` is larger than what is found necessary by the model, the
 Variational Bayesian mixture model has a natural tendency to set some mixture
 weights values close to zero. This makes it possible to let the model choose a
 suitable number of effective components automatically.
```
The formatting of this paragraph is broken: indentation is missing.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
+Here, a classical Gaussian mixture is fitted with 5 components on a dataset
+composed of 2 clusters. We can see that the Dirichlet Gaussian mixture is able
+to limit itself to only 2 components whereas the Gaussian mixture fits the data
+fit with too many components. Note that with very little observations, the
```
=> whereas the Gaussian mixture fits the data with a fixed number of components that has to be set a priori by the user. In this case the user has select `n_components=5` which does not match the true generative distribution of this toy dataset.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,75 @@ Cons
+   :align: center
+   :scale: 70%
+
+Here, a classical Gaussian mixture is fitted with 5 components on a dataset
```
I think the comma is superfluous.

ogrisel commented:
In the description #1637 is said to be fixed by this PR but I cannot find any test w.r.t. to sampling nor can I find a sample method on the new classes. I think this is missing.

ogrisel commented:
I think we also need to extend one of the examples to demonstrate sampling. Maybe it would be interesting to add a new panel to the "sin" example to do a scatter plot of points samples from the last Dirichlet process model.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -93,14 +93,14 @@ Cons
-number of components for a Gaussian mixture model.
+The BIC criterion can be used to select the number of components in a Gaussian
+Mixture in an efficient way. In theory, it recovers the true number of
+components only in the asymptotic regime (i.e. if much data is available). Note
```
(i.e. if much data is available and assuming that the data was actually generated i.i.d. from a mixture of Gaussian distribution)

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,78 @@ Cons
 
-.. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
+
+The examples bellow compare Gaussian mixture models with fixed number of
```
...with a fixed number...

tguillemot commented:
> In the description #1637 is said to be fixed by this PR but I cannot find any test w.r.t. to sampling nor can I find a sample method on the new classes. I think this is missing.

I'm not sure it's really important to have a `sample` function for the GMM. Numpy gives `multivariate_normal` which does that perfectly. 

When I've put that fix it was to close directly the old issue by merging this PR. 
It was a mistake. I correct that.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,78 @@ Cons
 
-.. centered:: |plot_gmm| |plot_gmm_sin|
+On the following example, we are fitting a dataset not well-depicted by a Gaussian
+mixture. Adjusting the ``weight_concentration_prior``, parameter of the DPGMM
```
Please remove the comma.

tguillemot committed:
[ci -skip] Fix typos.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,78 @@ Cons
+The next figure presents the resulting clusters computed by the Gaussian mixture
+with a Dirichlet process prior for different values of
+``weight_concentration_prior``. The ``weight_concentration_prior`` is related to
+the number of clusters obtained.  As in Variational Bayesian Gaussian Mixture
```
The ``weight_concentration_prior`` is related to the number of clusters obtained. 

=>

The value of the ``weight_concentration_prior`` parameter has a strong impact on the effective number of active components obtained.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +211,78 @@ Cons
+with a Dirichlet process prior for different values of
+``weight_concentration_prior``. The ``weight_concentration_prior`` is related to
+the number of clusters obtained.  As in Variational Bayesian Gaussian Mixture
+(for the ``weight_concentration_prior`` parameter), smaller values of
```
Please remove "(for the ``weight_concentration_prior`` parameter)" which is redundant now that the two parameters have the same name.

agramfort replied:
I think sample method is useful to have a generative model that can be used for a classifier eg to model p(x|y)
It does more than scipy






On Tue, Sep 6, 2016 at 4:33 PM +0200, "Olivier Grisel" <notifications@github.com> wrote:












In doc/modules/mixture.rst:
>  
> +.. figure:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_sin_001.png
> +   :target: ../auto_examples/mixture/plot_gmm_sin.html
> +   :align: center
> +   :scale: 65%
> +
> +The next figure presents the resulting clusters computed by the Gaussian mixture
> +with a Dirichlet process prior for different values of
> +``weight_concentration_prior``. The ``weight_concentration_prior`` is related to
> +the number of clusters obtained.  As in Variational Bayesian Gaussian Mixture
> +(for the ``weight_concentration_prior`` parameter), smaller values of



Please remove "(for the weight_concentration_prior parameter)" which is redundant now that the two parameters have the same name.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -265,10 +294,11 @@ Pros
+number needs to be provided. Note however that the "ideal" number of active
+components is very application specific and is typically ill-defined in a data
+exploration setting. The effective number of active components often depends a
+lot on the value for the weights concentration prior.
```
The indentation is broken here.

tguillemot commented:
> I think sample method is useful to have a generative model that can be used for a classifier eg to model p(x|y). It does more than scipy

Ok let's do it.

ogrisel commented:
> Numpy gives multivariate_normal which does that perfectly. 

No numpy does not sample from a mixture but from a single component model. Please re-add (and maybe adapt) the old `GMM.sample` method into the new mixture base class.

tguillemot commented:
> No numpy does not sample from a mixture but from a single component model. Please re-add (and maybe adapt) the old GMM.sample method into the new mixture base class.

Indeed sorry for the mistake.

tguillemot committed:
Add sample function for all mixture classes + fix multivariate_normal issue.

tguillemot commented:
Ok I've added the sampling function.

tguillemot commented:
Here is the resulting sampling :
![sampling](https://cloud.githubusercontent.com/assets/17002898/18305952/024812d4-74ea-11e6-893e-96fb5939e421.png)

I've added the what's new and also some comments on the deprecated files to explain how to clean 0.20.

TomDLT commented on examples/mixture/plot_gmm_sin.py:
```diff
@@ -79,20 +97,31 @@ def plot_results(X, Y_, means, covariances, index, title):
+# Sample the Dirichlet process Gaussian mixture
+X_s, y_s = dpgmm.sample(n_samples=5000)
+plot_samples(X_s, y_s, dpgmm.n_components, 0, "Sampling of model obtained "
+             r"with a Dirichlet Process with $\beta_0=100$ and $5000$ samples")
```
$\beta_0=0.01$

tguillemot replied:
Thanks

tguillemot committed:
[ci skip] Fix typo.

agramfort commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +213,80 @@ Cons
 
-.. |plot_gmm| image:: ../auto_examples/mixture/images/sphx_glr_plot_gmm_001.png
+
+The examples bellow compare Gaussian mixture models with a fixed number of
```
bellow -> below

agramfort commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +213,80 @@ Cons
+of 2 clusters. We can see that the Gaussian mixture with a Dirichlet process
+prior is able to limit itself to only 2 components whereas the Gaussian mixture
+fits the data with a fixed number of components that has to be set a priori by
+the user. In this case the user has select ``n_components=5`` which does not
```
select -> selected

agramfort commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,124 @@
+parameter ``weight_concentration_prior`` has a direct link with the resulting
+number of components. Specifying higher values more often leads to uniformly-
+sized mixture components, while specifying smaller values will lead to some
+mixture components getting almost all the points while most of the other
```
while -> with

agramfort commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,124 @@
+mixture components getting almost all the points while most of the other
+mixture components centering on just a few of the remaining points.
+
+The Dirichlet Process allows to definite an infinite number of component and
```
definite -> define

agramfort replied:
number of component -> number of components

agramfort commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,124 @@
+
+The Dirichlet Process allows to definite an infinite number of component and
+automatically selects the correct number of components. Contrary to the
+classical variation Bayesian model, it activates a components only if it is
```
components -> component

agramfort commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,124 @@
+The Dirichlet Process allows to definite an infinite number of component and
+automatically selects the correct number of components. Contrary to the
+classical variation Bayesian model, it activates a components only if it is
+necessary (resulting a better selection of the mixtures).
```
resulting -> resulting in

agramfort commented on examples/mixture/plot_gmm_sin.py:
```diff
@@ -75,16 +97,31 @@ def plot_results(X, Y_, means, covariances, index, title):
+plot_samples(X_s, y_s, dpgmm.n_components, 0, "Sampling of model obtained "
+             r"with a Dirichlet Process with $\beta_0=0.01$ and $5000$ samples")
+
+# Fit a Gaussian mixture models with a Dirichlet process prior ten components
```
ten components -> and ten components

agramfort commented on sklearn/mixture/base.py:
```diff
@@ -355,6 +355,51 @@ def predict_proba(self, X):
+
+        Returns
+        -------
+        X : array, shape (n_features, n_samples)
```
should be (n_samples, n_features)

agramfort commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -251,23 +254,26 @@ class BayesianGaussianMixture(BaseMixture):
     --------
     GaussianMixture : Finite Gaussian mixture fit with EM.
+
+    DirichletGaussianMixture : An infinite gaussian mixture model fit with a
```
gaussian -> Gaussian

agramfort commented on sklearn/mixture/dirichlet_gaussian_mixture.py:
```diff
@@ -0,0 +1,338 @@
+    a fixed maximum number of components, but almost always the number
+    of components actually used depends on the data.
+
+    Stick-breaking Representation of a Gaussian mixture model
```
Stick-breaking Representation of a Gaussian mixture model
 +    probability distribution

this is not a sentence

agramfort commented on sklearn/mixture/dirichlet_gaussian_mixture.py:
```diff
@@ -0,0 +1,338 @@
+    mean_precision_prior : float, optional.
+        The user-provided mean precision prior parameter of the Gaussian
+        distribution. Controls the extend to where means can be placed. Smaller
+        values concentrates the means of each clusters around `mean_prior`.
```
concentrates -> concentrate

agramfort commented:
that's it for me.

then +1 for merge

tguillemot committed:
Fix typos.

agramfort commented:
travis is not happy but besides +1 for merge

tguillemot commented:
Master is broken.

jnothman commented:
oh, I did that... whoops.

jnothman commented:
Big apologies. I apparently hurried a merge and then included unwanted code in a fix to it.

tguillemot commented:
@jnothman No problem it can happen ;)

ogrisel commented:
By re-reading the code, the doc and the examples, I think our class hierarchy and naming is unnecessarily complex and confusing:

I think the `BayesianGaussianMixture` and the `DirichletGaussianMixture` classes should be merged into a single class (also named `BayesianGaussianMixture`) and we should introduce a new parameter to switch between the 2:

- `weight_concentration_prior_type` with values in `"dirichlet_process"` (to enable the code that currently lives in the `DirichletGaussianMixture` class and `"dirichlet_distribution"` (to enabled the code that currently lives in `BayesianGaussianMixture`.

- I also think we should have `concentration_prior_type="dirichlet_process"` by default as in all the examples it seems to display a more intuitive behavior w.r.t. to the selection of active components.

- We might also want rename `weight_concentration_prior{_type}` to just `concentration_prior{_type}` to make the parameter names shorter while still being explicit enough (I don't think they can be confused with any other parameter in this case).

The documentation would be much simpler in my opinion.

What other people think? @tguillemot @xuewei4d @agramfort @jnothman

@amueller as this change would impact the API quite significantly I would rather have it agreed upon before the release.

ogrisel commented:
BTW thanks @tguillemot for addressing the comments so quickly, this is really appreciated.

tguillemot commented:
+1 for the API simplification.

agramfort commented:
brilliant idea @ogrisel !

ogrisel commented:
Ok that's 3 people who like the idea of merging the two variational estimators into a single class. If nobody objects, I suggest @tguillemot you can do the merge as part of this PR.

ogrisel commented:
Small issue, when `max_iter` is too small one gets the following message:

> ConvergenceWarning: Initialization 1 did not converged. Try different init parameters, or increase n_init, tol or check for degenerate data.

This should be `max_iter` instead of `n_init`.

ogrisel commented:
BTW I confirm that this model fixes the scaling bug reported initially for DGPMM: #2454: the estimated means for the components with large weights are reasonably close to the ground truth means. Is there a test for this?

tguillemot committed:
Change the Api by merging DirichletGaussianMixture into BayesianGaussianMixture.

Add the parameter `weight_concentration_prior_type`to select between the Dirichlet
process and the Dirichlet distribution to model the BayesianGaussianMixture.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -189,11 +210,11 @@ Pros
-   weights values close to zero. This makes it possible to let the model choose a
-   suitable number of effective components automatically.
+   weights values close to zero. This makes it possible to let the model choose
+   a suitable number of effective components automatically.
```
Please remove / merge the pros and the cons section with the one for the Dirichlet process variant.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -211,49 +232,67 @@ Cons
-DPGMM: Infinite Gaussian mixtures
-=================================
+Dirichlet Process: Infinite Gaussian mixture
+============================================
```
Please make this a sub-section of "Variational Bayesian Gaussian Mixture" by using "-------" underline.

ogrisel replied:
Actually I would be in favor of removing this subsection completely and merging the content with the previous section to remove redundancies.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -152,35 +154,54 @@ priors) instead of data likelihood. The principle behind
+Due to its Bayesian nature, the variational algorithm needs more hyper-
+parameters than expectation-maximization, the most important of these being the
+concentration parameter ``weight_concentration_prior``. Specifying a high value
+of prior of the dirichlet concentration leads more often to uniformly-sized
```
"leads more often to uniformly-sized" this is only true for the Dirichlet distribution prior. Instead I would state:

Specifying a low value for the concentration prior will make the model put most of the weight on few components set the remaining components weights very close to zero. High values of the concentration prior will allow a larger number of components to be active in the mixture.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -152,35 +154,54 @@ priors) instead of data likelihood. The principle behind
+``weight_concentration_prior``. Here, we can see the the value of the
+``weight_concentration_prior`` parameter has a strong impact on the effective
+number of active components obtained.
+
```
We can also notice that large values for the concentration weight prior lead to more uniform weights when the type of prior is `'dirichlet_distribution'` while this is not necessarily the case for the `'dirichlet_process'` type (used by default).

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,130 @@
+
+This example plots the ellipsoids obtained from a toy dataset (mixture of three
+Gaussians) fitted by the ``BayesianGaussianMixture`` class models with a
+Dirichlet process distribution
```
Dirichlet distribution prior

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,130 @@
+Gaussians) fitted by the ``BayesianGaussianMixture`` class models with a
+Dirichlet process distribution
+(``weight_concencration_prior_type='dirichlet_distribution'``) and a Dirichlet
+process prior (``weight_concencration_prior_type='dirichlet_process'``). On
```
concentration

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,130 @@
+each figure, we plot the results for three different values of the weight
+concentration prior.
+
+The ``BayesianGaussianMixture`` can adapt its number of mixture automatically.
```
The ``BayesianGaussianMixture`` class

ogrisel replied:
its number of mixture components

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,130 @@
+
+The ``BayesianGaussianMixture`` can adapt its number of mixture automatically.
+The parameter ``weight_concentration_prior`` has a direct link with the
+resulting number of components. Specifying higher values more often leads to
```
resulting number of components with non-zero weights.

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,130 @@
+The ``BayesianGaussianMixture`` can adapt its number of mixture automatically.
+The parameter ``weight_concentration_prior`` has a direct link with the
+resulting number of components. Specifying higher values more often leads to
+uniformly- sized mixture components, while specifying smaller values will lead
```
Please remove the reference to uniformly distributed components which is not true for the DP prior.

Add a note at the end of the paragraph stating that the DD prior will favor more uniformly weighted components.

ogrisel commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -63,30 +63,38 @@ def _log_wishart_norm(degrees_of_freedom, log_det_precisions_chol, n_features):
     parameters of a Gaussian mixture distribution. The effective number of
     components can be inferred from the data.
 
+    This class implements two different prior type for the weight distribution:
```
types of prior

ogrisel commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -63,30 +63,38 @@ def _log_wishart_norm(degrees_of_freedom, log_det_precisions_chol, n_features):
+    model with the Dirichlet Process. In practice the approximate the Dirichlet
+    Process inference algorithm uses a truncated distribution with a fixed
+    maximum number of components (called the Stick-breaking representation),
+    but almost always the number of components actually used depends on the
```
Cut the sentence:

representation). The number of components actually used almost always depends on the data.

ogrisel commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -63,30 +63,38 @@ def _log_wishart_norm(degrees_of_freedom, log_det_precisions_chol, n_features):
+    This class implements two different prior type for the weight distribution:
+    a finite mixture model with Dirichlet distribution and an infinite mixture
+    model with the Dirichlet Process. In practice the approximate the Dirichlet
+    Process inference algorithm uses a truncated distribution with a fixed
```
In practice Dirichlet Process inference algorithm is approximated and uses a truncated distribution with a fixed...

ogrisel commented:
CI is broken because of a broken import and a PEP8 issue.

Also could you please try to make the following test run in less than 1s by tweaking the training data size or the hyper parameters of the model?

```
sklearn.mixture.tests.test_bayesian_mixture.test_monotonic_likelihood: 4.5968s
```

tguillemot committed:
Add tests + change mixture doc.

tguillemot commented:
Thanks @ogrisel.
I've changed the doc and add the missing tests.

ogrisel commented on doc/modules/mixture.rst:
```diff
@@ -152,171 +155,167 @@ priors) instead of data likelihood. The principle behind
+      :class:`GaussianMixture` and :class:`BayesianGaussianMixture` to fit a
+      sine wave.
+
+    * See :ref:`sphx_glr_auto_example_mixture_plot_concentration_prior.py`
```
This reference is broken.

tguillemot commented:
I have a problem of the a ghost file in travis. @ogrisel can you remove the cache ?

ogrisel commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -111,7 +118,14 @@ class BayesianGaussianMixture(BaseMixture):
 
-    dirichlet_concentration_prior : float | None, optional.
+    weight_concentration_prior_type : {'dirichlet_process',
+            'dirichlet_distribution'}, defaults to 'full'.
```
Two problems:

- This new line breaks the sphinx rendering in classes.rst. I am not sure how this should be fixed.
- The default value is `'dirichlet_process'`, not `'full'`.

tguillemot replied:
I don't know how to fix the first problem because if I put it on a unique line pep8 will not be happy.

ogrisel commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -63,30 +63,37 @@ def _log_wishart_norm(degrees_of_freedom, log_det_precisions_chol, n_features):
+    The number of components actually used almost always depends on the data.
+
     Read more in the :ref:`User Guide <bgmm>`.
 
```
Please insert the following marker:

    .. versionadded:: 0.18
       *BayesianGaussianMixture*.

A similar marker should be inserted for the `GaussianMixture` class.

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,131 @@
+
+# mean_precision_prior= 0.8 to minimize the influence of the prior
+estimators = [
+    (r"Bayesian Gaussian Mixture for $\gamma_0=$", BayesianGaussianMixture(
```
Please change the title with "Finite mixture with Dirichlet distribution prior with $\gamma_0=$"" (maybe with a line break `\n` after "prior".

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,131 @@
+        n_components=2 * n_components, reg_covar=0, init_params='random',
+        max_iter=1500, mean_precision_prior=.8,
+        random_state=random_state), [0.001, 1, 1000]),
+    (r"Dirichlet Process Mixture for $\gamma_0=$", BayesianGaussianMixture(
```
Please change the title with "Infinite mixture with a Dirichlet process prior and $\gamma_0=$"" (maybe with a line break `\n` after "prior".

ogrisel commented on examples/mixture/plot_concentration_prior.py:
```diff
@@ -0,0 +1,131 @@
+automatically selects the correct number of components. Contrary to the
+classical variation Bayesian model using a Dirichlet distribution, it activates
+a component only if it is necessary (resulting in a better selection of the
+mixtures) and will favor more uniformly weighted components.
```
This is confusing, I would instead write:

...automatically selects the correct number of components: it activates a component only if it is necessary.

On the contrary the classical finite mixture model with a Dirichlet distribution prior will favor more uniformly weighted components and therefore tends to divide natural clusters into unnecessary sub-components.

ogrisel commented:
I think the red travis builds were caused by old cached versions of deleted python modules and test files. I manually deleted the cache for this PR in travis and relaunched the build to check if that fixes it.

ogrisel commented:
pyflakes has caught an unused variable:

    ./sklearn/mixture/tests/test_bayesian_mixture.py:403:19: F841 local variable 'n_features' is assigned to but never used
        n_components, n_features = 2 * rand_data.n_components, 2

ogrisel commented on examples/mixture/plot_gmm_sin.py:
```diff
@@ -3,15 +3,16 @@
+process prior, however, show that the model can either learn a global structure
+for the data (small ``weight_concentration_prior``) or easily interpolate to
+finding relevant local structure (large ``weight_concentration_prior``), never
+falling into the problems shown by the ``GaussianMixture`` class.
```
I don't agree with this analysis. Let me suggest the following instead:

This example demonstrates the behavior of Gaussian mixture models on data that was not generated by a mixture of Gaussian random variables.  The dataset is formed by 100 points loosely spaced following a noisy sine curve.  There is therefore no ground truth value for the number of Gaussian components.

The first model is a classical Gaussian Mixture Model with 10 components fit with the Expectation Maximization algorithm.

The second model is a Bayesian Gaussian Mixture Model  with a Dirichlet process prior fit with variational inference. The low value of the concentration prior makes the model favor a lower number of active components. This models "decides" to focus its model power on the big picture of the structure of the datasets: groups of points with alternating directions modeled by non-spherical covariance matrices. Those alternating directions roughly capture the alternating nature of the original sine signal.

The third model is also Bayesian Gaussian Mixture Model  with a Dirichlet process prior but this time the value of the concentration prior is higher giving the model more liberty to try to model the finer-grained structure of the data. The result is a mixture with a larger number of active components that is similar to the first model where we decided to fix the number of components to 10 arbitrarily.

Which models is best is a matter of subjective judgement:  do we want to favor models that only capture the big picture to summarize and explain most of the structure of the data while ignoring the details or do we prefer models that closely follow the high density regions of the signal?

The last two panels show how we can sample from the last two models. The resulting samples distributions do not look exactly like the original data distribution. The difference primarily stems from the approximation error we made by using a model that assumes that the data was generated by a finite number of Gaussian components instead of a noisy sine curve.

ogrisel commented:
This time the travis error is for real I think:

https://travis-ci.org/scikit-learn/scikit-learn/jobs/158774349#L2336

ogrisel commented:
I think I am done with the review. +1 for merge once CI is green and my last comment on `plot_gmm_sin.py` has been taken into account. Thanks very much for bearing with me @tguillemot :)

ogrisel commented:
I fixed the CI failure and addressed the doc of the example in #7386. If it's green, I will merge.

ogrisel commented:
Merged as #7386 :beers:

tguillemot commented:
@ogrisel Sorry I had to go yesterday. Thanks for taking care of that.

tguillemot commented:
Thanks everyone for your review and helps !!!
